controller/snapshot_controller.go

```diff
@@ -344,8 +347,25 @@ func (sc *SnapshotController) reconcile(snapshotName string) (err error) {
 			return sc.handleAttachmentTicketDeletion(snapshot)
 		}
 
-		if err := sc.handleAttachmentTicketCreation(snapshot); err != nil {
-			return err
+		// It is possible for us to be reconciling an out-of-date snapshot that has had its finalizer removed and has
+		// been deleted from the cluster, so we must take special care around attachment tickets.
+		if val, ok := snapshot.Annotations[types.AttachingForDeletionAnnotationKey]; ok && val == "true" {
+			// It is fine to update an attachment ticket more than once after DeletionTimestamp is set. If we attempt
+			// to update an out-of-date volume attachment, we will simply fail.
+			if err := sc.handleAttachmentTicketModification(snapshot, false); err != nil {
+				return err
+			}
+		} else {
+			// Only create an attachment ticket once after DeletionTimestamp is set and then immediately return. We must
+			// not create an attachment ticket again in case this is the last time we reconcile.
+			if err := sc.handleAttachmentTicketModification(snapshot, true); err != nil {
+				return err
+			}
+			if snapshot.Annotations == nil {
+				snapshot.Annotations = make(map[string]string)
+			}
+			snapshot.Annotations[types.AttachingForDeletionAnnotationKey] = "true"
+			return nil
```

Comment:
```suggest
requireVATicket := snapshot.Annotations[types.AttachingForDeletionAnnotationKey] == "true"
if err := sc.handleAttachmentTicketModification(snapshot, requireVATicket); err != nil {
	return err
}
snapshot.Annotations[types.AttachingForDeletionAnnotationKey] = "true"
```

Reply:
BTW, we can do `snapshot.Annotations = make(map[string]string)` in the mutation webhook.

Reply:
Good suggestions. They clean things up a bit since:

- Checking for `ok` and `"true"` is redundant. If `!ok`, then we get `""`.
- In both cases, we take the same basic action. The only differente is the boolean we pass.
- It is fine to set the annotation to true in either case (even if it's been done before).
- We do not actually need to automatically return.

> BTW, we can do snapshot.Annotations = make(map[string]string) in the mutation webhook.

Yes, this can be done, but I prefer we do a nil check every time we start writing to a map, just in case.